I'm going to add a note to this, as I don't understand why the definition of `WM_LBUTTONDOWN` in `/usr/local/include/freerdp/rail.h` is not triggering a macro replacement error.

Consider these files:-

```
// temp.h
#define WM_LBUTTONDOWN 0x202
```

```
// temp.c
#define WM_LBUTTONDOWN 102
#include <temp.h>

int main()
{
    return 0;
}
```

Then :-

```
$ gcc -I. -g temp.c
In file included from temp.c:4:
./temp.h:1: warning: "WM_LBUTTONDOWN" redefined
    1 | #define WM_LBUTTONDOWN 0x202
      | 
temp.c:1: note: this is the location of the previous definition
    1 | #define WM_LBUTTONDOWN 102
      | 
```

However, if I move temp.h to the compiler's default include path:-

```
sudo cp temp.h /usr/local/include
gcc -g temp.c
```

I don't get any errors, despite the macro definition still being accepted. The replacement is happening silently:-

```
gcc -E -dM temp.c | grep WM_LBUTTONDOWN
#define WM_LBUTTONDOWN 0x202
```

I assume this is some gcc specialisation - clang does the same thing so I'm assuming it's not a bug. Anyone got any thoughts?

--- a/neutrinordp/xrdp-color.c
+++ b/neutrinordp/xrdp-color.c
@@ -22,6 +22,10 @@
 #endif
 
 #include "xrdp-neutrinordp.h"
+#include "defines.h"
+#include "log.h"
+#include "os_calls.h"
+#include "string_calls.h"
 
 char *
 convert_bitmap(int in_bpp, int out_bpp, char *bmpdata,

--- a/neutrinordp/xrdp-neutrinordp.c
+++ b/neutrinordp/xrdp-neutrinordp.c
@@ -17,6 +17,17 @@
  * limitations under the License.
  */
 
+#include <freerdp/freerdp.h>
+#include <freerdp/codec/bitmap.h>
+
+// FreeRDP defines some macros that we have different values for.
+// To catch this we need to include the freerdp includes before our
+// local ones (see gcc bug #16358)
+#undef WM_LBUTTONUP
+#undef WM_LBUTTONDOWN
+#undef WM_RBUTTONUP
+#undef WM_RBUTTONDOWN
+
 #if defined(HAVE_CONFIG_H)
 #include <config_ac.h>
 #endif
@@ -26,8 +37,8 @@
 #include "xrdp_rail.h"
 #include "trans.h"
 #include "log.h"
+#include "os_calls.h"
 #include "string_calls.h"
-#include <freerdp/settings.h>
 
 #if defined(VERSION_STRUCT_RDP_FREERDP)
 #if VERSION_STRUCT_RDP_FREERDP > 1

--- a/neutrinordp/xrdp-neutrinordp.h
+++ b/neutrinordp/xrdp-neutrinordp.h
@@ -20,22 +20,15 @@
 #ifndef XRDP_NEUTRINORDP_H
 #define XRDP_NEUTRINORDP_H
 
-/* include other h files */
 #include "arch.h"
-#include "parse.h"
-#include "os_calls.h"
-#include "defines.h"
-#include "xrdp_rail.h"
-#include "xrdp_client_info.h"
 #include "xrdp_constants.h"
+#include "xrdp_client_info.h"
 
-/* this is the freerdp main header */
-#include <freerdp/freerdp.h>
-#include <freerdp/rail.h>
-#include <freerdp/rail/rail.h>
-#include <freerdp/codec/bitmap.h>
-//#include <freerdp/utils/memory.h>
-//#include "/home/jay/git/jsorg71/staging/include/freerdp/freerdp.h"
+/* Incomplete type definitions, referenced below */
+struct rail_window_state_order;
+struct rail_notify_state_order;
+struct rail_monitored_desktop_order;
+struct rail_icon_info;
 
 struct bitmap_item
 {

